### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by ipanr-Nuba**

### CODE REVIEW

File: Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package com.example.isaacspetapp.views.upload
 
+import `drawable-v24`.checkForCameraPermissions
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: This is related to the wrong package in a previous comment.

File: Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt [hunk 2]
```diff
@@ -30,36 +30,38 @@ class UploadImageFragment: Fragment() {
         container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View? {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: The function declaration is 116 characters long if put in a single line. Which still fit the 120 limit.

File: Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt [hunk 3]
```diff
@@ -30,36 +30,38 @@ class UploadImageFragment: Fragment() {
...
-        val view = inflater.inflate(R.layout.fragment_uploadimage, container, false)
+        val view = inflater.inflate(com.example.isaacspetapp.R.layout.fragment_uploadimage,
+            container, false)
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: No reason to have this fragment in a new line. Still under 120 as a single line.

File: Project/app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -1,10 +1,11 @@
...
     <string name="app_name">IsaacsPetApp</string>
 
-    <!-- TODO: Remove or change this placeholder text -->
     <string name="hello_blank_fragment">Hello blank fragment</string>
     <string name="DollyText">Dolly Text</string>
     <string name="btn_txt_Upload_Image">Upload Image</string>
     <string name="editText_enterImageName">Enter image name</string>
-    <string name="btn_select_image">Select Image</string>
+    <string name="btn_select_image">Take photo</string>
+    <string name="Toast_NoNameOrPhoto">Please, add a photo and name it in order to upload it</string>
```
Comment (osalinasv) [inline Project/app/src/main/res/values/strings.xml]: String resource name does not match standard

```suggestion
    <string name="toast_no_name_or_photo">Please, add a photo and name it in order to upload it</string>
```

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt [hunk 5]
```diff
@@ -1,28 +1,73 @@
...
+
+    private lateinit var tempImageURI : Uri
+
+
+    fun takePhoto(context: Context): Intent {
+        val timeStamp: String = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
+        val timeStampString = "IMG_${timeStamp}"
+        var contentValues = ContentValues()
+        contentValues.put(MediaStore.Images.Media.TITLE, timeStampString)
+        tempImageURI = context.contentResolver?.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, contentValues)!!
```
[the remark this answers is not in the document]
Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: ok! I tought that the idea of that architecture was to arange the data; but that makes more sense. 
I will modify it.
Then it is ok if the view implements some methods that return data?

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt [hunk 6]
```diff
@@ -1,28 +1,73 @@
...
     private var _imageURI = MutableLiveData<Uri>()
-    val imageURL: LiveData<Uri>
+    val imageURI: LiveData<Uri>
         get() = _imageURI
-    private var _imageName = MutableLiveData<String>()
-    val imageName: LiveData<String>
-        get() = _imageName
+    private val IMAGE_CAPTURE_CODE = 10
+    val image_capture_code: Int
+        get() = IMAGE_CAPTURE_CODE
```
[the remark this answers is not in the document]
Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: I will look for it; i do not really know the usage of companion objects, but will look for it

File: Project/app/src/main/java/com/example/isaacspetapp/interfaces/ImgurInterfaceAPI.kt [hunk 7]
```diff
@@ -2,14 +2,21 @@ package com.example.isaacspetapp.interfaces
...
     @Headers("Authorization: Client-ID f6504eaf14f266b")
-    @GET("album/t4LliPu")
+    @GET("album/tcIjmuS")
     fun getAlbum(): Call<DataContainer<AlbumData>>
 
+    @FormUrlEncoded
+    @Headers("Authorization: Client-ID f6504eaf14f266b")
+    @POST("upload")
+    fun uploadPhoto(@Field("image") base64: String,
+                    @Field("album") album: String,
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/interfaces/ImgurInterfaceAPI.kt]: Parameters are usually aligned one tab level after the function declaration and starting on a newline, not in the declaration line. With the closing parenthesis and return type also on its own line.

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/CheckPermissions.kt [hunk 8]
```diff
@@ -8,22 +8,23 @@ import androidx.core.app.ActivityCompat
 import androidx.core.content.ContextCompat
 import androidx.fragment.app.FragmentActivity
 
-fun checkForImagePermissions(activity: FragmentActivity, context: Context): Boolean {
+fun checkForCameraPermissions(activity: FragmentActivity, context: Context): Boolean {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/CheckPermissions.kt]: If this function depends on a fragment/activity and a context, why not make it an extension of Fragment?

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+fun checkForEmptyTitles(imageInfo: ImageInfo): String {
+    var name = R.string.dollyText.toString()
+
+    if (imageInfo.title != null) {
+        name = imageInfo.title
+    }
+
+    return name
+}
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt]: Missing EOF newline.

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/convertTime.kt [hunk 10]
```diff
@@ -4,7 +4,6 @@ import java.text.SimpleDateFormat
 import java.util.*
 
 fun convertTime(unixTime: Long): String {
-    var df = java.util.Date(unixTime*1000)
-    val vv = SimpleDateFormat("MM dd, yyyy hh:mma", Locale.US).format(df)
-    return vv
-}
\ No newline at end of file
+    val timeToConvert = Date(unixTime*1000)
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/convertTime.kt]: Missing spaces before and after operator `*`.

File: Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt [hunk 11]
```diff
@@ -1,65 +1,107 @@
...
+    }
+
+    private fun observeViewModel() {
+        viewModel.imageURI.observe(this, Observer {uri ->
+            uf_image_imagetoupload.setImageURI(uri)
+        })
+
+        viewModel.uploadSuccessful.observe(this, Observer { uploadSuccess ->
+            if (uploadSuccess) {
+             Toast.makeText(context, R.string.upload_Successful, Toast.LENGTH_LONG).show()
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: Fix this line's indentation.

File: Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt [hunk 12]
```diff
@@ -1,65 +1,107 @@
...
+        val intent =Intent(MediaStore.ACTION_IMAGE_CAPTURE)
+        intent.putExtra(MediaStore.EXTRA_OUTPUT, tempImageURI)
+        startActivityForResult(intent, viewModel.imageCaptureCode)
     }
 
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         super.onActivityResult(requestCode, resultCode, data)
-        if (resultCode == Activity.RESULT_OK && requestCode == viewModel.image_pick_code) {
-            data?.data?.let { viewModel.setImageURIValue(it) }
+        if(resultCode == Activity.RESULT_OK) {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: Missing space after if.

File: Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt [hunk 13]
```diff
@@ -1,65 +1,107 @@
...
+        contentValues.put(MediaStore.Images.Media.TITLE, nameWithTimeStamp)
+        tempImageURI = this.context?.contentResolver?.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, contentValues)!!
+        val intent =Intent(MediaStore.ACTION_IMAGE_CAPTURE)
+        intent.putExtra(MediaStore.EXTRA_OUTPUT, tempImageURI)
+        startActivityForResult(intent, viewModel.imageCaptureCode)
     }
 
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         super.onActivityResult(requestCode, resultCode, data)
-        if (resultCode == Activity.RESULT_OK && requestCode == viewModel.image_pick_code) {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: You had the space before, but you removed it?

Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: not sure what happened there

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/CheckPermissions.kt (hunk 8 above)
Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/utilities/CheckPermissions.kt]: Will look for it!

File: Project/app/build.gradle [hunk 14]
```diff
@@ -26,6 +26,17 @@ android {
...
+    compileOptions {
+        sourceCompatibility = 1.8
+        targetCompatibility = 1.8
+    }
+
+    kotlinOptions {
+        jvmTarget = "1.8"
+    }
+    testOptions {
+        unitTests.returnDefaultValues = true
```
Comment (osalinasv) [inline Project/app/build.gradle]: In general it's not recommended to have `returnDefaultValues` set to true. It might hide errors in the tests or make it more difficult to track problems.

```suggestion
        unitTests.returnDefaultValues = false
```

File: Project/app/src/main/java/com/example/isaacspetapp/MainApplication.kt [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    override val kodein = Kodein.lazy {
+        import(androidXModule(this@MainApplication))
+
+        bind<APIImgurServices>() with singleton { APIImgurServicesImp() }
+        bind<OkHttpServices>() with singleton { OkHttpServicesImp() }
+
+        bind() from provider { MainViewModelFactory(instance()) }
+        bind() from provider { UploadImageViewModelFactory(instance()) }
+        bind() from provider { DetailsViewModelFactory(instance()) }
+
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/MainApplication.kt]: Remove extra newline

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt [hunk 16]
```diff
@@ -0,0 +1,24 @@
+package com.example.isaacspetapp.utilities
+
+import com.example.isaacspetapp.R
+import com.example.isaacspetapp.models.ImageInfo
+
+fun checkForEmptyTitles(imageInfo: ImageInfo): String {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt]: The naming in this function is confusing. The word check might be interpreted as a boolean check, so it would mean that the function tests if the "title" is empty. But it looks like the function returns either the title if it exists or a default text.

Also, if the only property of `imageInfo` that is used is the `title` String, why not pass the string as a parameter directly?

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    var name = R.string.dollyText.toString()
+
+    if (imageInfo.title != null) {
+        name = imageInfo.title
+    }
+
+    return name
+}
+
+fun checkForEmptyText(enteredValue: String?): String {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt]: Similar to the previous comment, the function does not really return a check for empty texts.

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+    var Text = UtilityStrings.dollyText
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt]: Variables should not be in PascalCase, only classes and interfaces.

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/HttpResponseCode.kt [hunk 19]
```diff
@@ -0,0 +1,17 @@
+package com.example.isaacspetapp.utilities
+
+class HttpResponseCode {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/HttpResponseCode.kt]: A collection of elements that share a single value type should be an enum

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/ImgurAPIEndpoints.kt [hunk 20]
```diff
@@ -0,0 +1,6 @@
+package com.example.isaacspetapp.utilities
+
+object ImgurAPIEndpoints {
+    val _albumDeleteHash = "9LsxWwMk5z5RtvG"
+    val _imgurApiBaseURL = "https://api.imgur.com/3/"
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/ImgurAPIEndpoints.kt]: You have this endpoint here but you do not use it in APIImgurServicesImp. You also do not save the album hash ID or the Client-ID, they are still hardcoded in ImgurInterfaceAPI.

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/UtilityStrings.kt [hunk 21]
```diff
@@ -0,0 +1,7 @@
+package com.example.isaacspetapp.utilities
+
+class UtilityStrings {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/UtilityStrings.kt]: Why have constant "utility" strings instead of using Android's string resources?
If you are going to have a static set of constants then use `object` instead of `class`

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt [hunk 22]
```diff
@@ -4,25 +4,90 @@ import android.net.Uri
...
+        if (enteredValue == null) return hasText
+
+        if (enteredValue.isNotEmpty()) {
+            _imageName = enteredValue
+            hasText = true
+        }
+        return hasText
+    }
+
+    private fun checkForExistingFile(baseValue: String?): Boolean {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: The naming of this function and its parameter is weird. How does checking if a string is not empty relate to checking if a File exists? And what does baseValue mean? What is it the base of?

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt [hunk 23]
```diff
@@ -4,25 +4,90 @@ import android.net.Uri
...
+                    _uploadSuccessful.postValue(true)
+                }
+            })
+
+        }
+
+        return canUpload
+    }
+
+    private fun checkForEmptyText(enteredValue: String?): Boolean {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: This makes much more sense than the other check-named functions I have commented. This does return the result of a check to tell wether the text is empty or not.

But at the same time you included side effects on the function, why do you set _imageName here? it seems unrelated to the purpose of the function.

This function can be greatly simplified.

File: Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt [hunk 24]
```diff
@@ -1,65 +1,157 @@
...
+            viewModel.setValuesOnResponse(tempImageURI)
+            viewModel.setImageBase64Values(getUriBytes(tempImageURI))
         }
     }
 
-    fun observeViewModel() {
-        viewModel.imageURL.observe(this, Observer {uri ->
-            uf_image_imagetoupload.setImageURI(uri)
-        })
+    private fun getUriBytes(uri: Uri): String {
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/views/upload/UploadImageFragment.kt]: The function is called getUriBytes but you get a base64 string representation of a Bitmap

File: Project/app/src/test/java/com/example/isaacspetapp/DetailsViewModelTest.kt [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+
+        viewModel.downloadPhoto(url, mockName)
+
+        viewModel.loadError.test().assertHasValue().assertValue(true)
+        viewModel.imageBitmap.test().assertNoValue()
+        viewModel.imageName.test().assertNoValue()
+    }
+
+    @Test
+    fun `downloadPhoto() empty Name ResponseCode OK`() {
```
Comment (osalinasv) [inline Project/app/src/test/java/com/example/isaacspetapp/DetailsViewModelTest.kt]: This test seems to be failing:
```
org.mockito.exceptions.misusing.InvalidUseOfMatchersException: 
Misplaced or misused argument matcher detected here:

-> at com.example.isaacspetapp.DetailsViewModelTest.any(DetailsViewModelTest.kt:51)

You cannot use argument matchers outside of verification or stubbing.
Examples of correct usage of argument matchers:
    when(mock.get(anyInt())).thenReturn(null);
    doThrow(new RuntimeException()).when(mock).someVoidMethod(anyObject());
    verify(mock).someMethod(contains("foo"))

This message may appear after an NullPointerException if the last matcher is returning an object 
like any() but the stubbed method signature expect a primitive argument, in this case,
use primitive alternatives.
    when(mock.get(any())); // bad use, will raise NPE
    when(mock.get(anyInt())); // correct usage use

Also, this error might show up because you use argument matchers with methods that cannot be mocked.
Following methods *cannot* be stubbed/verified: final/private/equals()/hashCode().
Mocking methods declared on non-public parent classes is not supported.


	at com.example.isaacspetapp.DetailsViewModelTest.setup(DetailsViewModelTest.kt:43)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

Comment (ipanr-Nuba) [reply Project/app/src/test/java/com/example/isaacspetapp/DetailsViewModelTest.kt]: This one os the only missing. I am soving it right now

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt [hunk 26]
```diff
@@ -0,0 +1,20 @@
+package com.example.isaacspetapp.utilities
+
+import com.example.isaacspetapp.R
+import com.example.isaacspetapp.models.ImageInfo
+
+fun returnDisplayImageNameIfNotEmpty(imageInfo: ImageInfo): String {
+    var name = R.string.dollyText.toString()
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt]: I still think the parameter should be a simple String instead of the whole ImageInfo if you are only going yo use the title property.

Also, you can save two allocations and a couple de-reference instructions if you return earlier:

```kotlin
return if (imageInfo.title != null) {
	imageInfo.title
} else {
	R.string.dollyText.toString()
}
```

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+
+    if (imageInfo.title != null) {
+        name = imageInfo.title
+    }
+
+    return name
+}
+
+fun returnUploadImageNameIfNotEmpty(enteredValue: String?): String {
+    if (enteredValue == null || enteredValue.isEmpty()) return UsefulStrings.dollyText
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/EmptyChecks.kt]: You can simplify this even more with null-coallesing.
You could also use `isNullOrBlank()` to also check for strings like `"  "`

```kotlin
return if (enteredValue.isNullOrEmpty()) UsefulStrings.dollyText else enteredValue
```

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/HttpResponseCode.kt [hunk 28]
```diff
@@ -0,0 +1,16 @@
...
+    const val ACCEPTED = 202
+    const val NO_CONTENT = 204
+    const val BAD_REQUEST = 400
+    const val UNAUTHORIZED = 401
+    const val FORBIDDEN = 403
+    const val NOT_FOUND = 404
+    const val RATE_LIMITING = 429
+    const val INTERNAL_SERVER_ERROR = 500
+    const val NOT_IMPLEMENTED = 501
+
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/HttpResponseCode.kt]: Remove newline. Also explain why you decided to not use an enum for this.

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/UsefulStrings.kt [hunk 29]
```diff
@@ -0,0 +1,5 @@
+package com.example.isaacspetapp.utilities
+
+object UsefulStrings {
+        const val dollyText = "Dolly Text"
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/utilities/UsefulStrings.kt]: What reason do you have to not use a string resource and to keep this hardcoded, untranslatable placeholder string here? Is the final user really going to see this in the UI?

Also the indentation is wrong by one level.

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt [hunk 30]
```diff
@@ -4,25 +4,85 @@ import android.net.Uri
...
+        if (enteredValue == null || enteredValue.isEmpty()) return false
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: Notice that you can return the boolean expression directly. It would only require to negate it as it returns false when the expression is true and vice versa.

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/UsefulStrings.kt (hunk 29 above)
Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/utilities/UsefulStrings.kt]: I was reading that in order to access R.String.anyString i need to pass a context to my function, and i dont think i should do that, because it is being used to assign a value after an image is downloaded in the detailsViewModel.

The string dollyText is placed in the offchance a photo is downloaded without a title. In that case, that string is returned as the title for the downloaded photo. I do not think the app allows us to upload photos without titles, but if an album has a photo that was not uploaded with the app and that photo does not have a title, then the app crashed when trying to assign the text title to display, so i added that check for safety, hope it's the right way to do it

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/HttpResponseCode.kt (hunk 28 above)
Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/utilities/HttpResponseCode.kt]: When i acces the CodeResponses, i map them to the int value defined there, and i do not think it can be done with enums without adding more code. For example, i use the responseCode OK in the tests to mock a good response back to a callback in UploadViewModelTest

            callback.onResponse(mockCall, Response.success(HttpResponseCode.OK, response))

with an enum, i think i would have to define the logic thas passes the value 200 when the responsecode is required. so that's why i left it that way

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt (hunk 30 above)
Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: i read this recomendation over and over but i can not figure out how to implement it :S i don't get how; i ended up leaving it as

    private fun checkForEmptyText(enteredValue: String?): Boolean { 
        return enteredValue.isNullOrEmpty()
    }

File: Project/app/src/main/java/com/example/isaacspetapp/utilities/UsefulStrings.kt (hunk 29 above)
Comment (osalinasv) [reply Project/app/src/main/java/com/example/isaacspetapp/utilities/UsefulStrings.kt]: Having default text strings is good. But they should be translatable resources, placeholders like this should only be used during development to test a concept. Never show a placeholder to the final user.

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt [hunk 31]
```diff
@@ -4,25 +4,82 @@ import android.net.Uri
...
+                }
+            })
+
+        }
+
+        return canUpload
+    }
+
+    private fun checkForEmptyText(enteredValue: String?): Boolean {
+        return enteredValue.isNullOrEmpty()
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: There's not much point in making a function that only uses another function over the only parameter. Instead just call the inner function instead of this one.

File: Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt [hunk 32]
```diff
@@ -4,25 +4,82 @@ import android.net.Uri
...
+    }
+
+    private fun checkForEmptyBase64Data(base64Value: String?): Boolean {
+        return base64Value.isNullOrEmpty()
```
Comment (osalinasv) [inline Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: There's not much point in making a function that only uses another function over the only parameter. Instead just call the inner function instead of this one.

Comment (ipanr-Nuba) [reply Project/app/src/main/java/com/example/isaacspetapp/viewmodels/UploadImageViewModel.kt]: all right!
